### ISSUE TITLE
Using avrgirl in Chrome app

### ISSUE THREAD

**Issue by joachimhs** [noopkat/avrgirl#3]
Hello, 

I am trying to integrate avrgirl into a Chrome app, but as avrgirl uses localStorage, the app wont load. Is there a way of fixing this issue without having to manually remove the calls to window.localStorage from the source?

**Comment by joachimhs** [comment]
I realise that this is coming in from a dependency. (a number of them, actually). I will investigate further. 

I've managed to get it working by commenting out the lines dealing with local storage, but this is not a viable option in the long run. 

On the positive side though it looks like I can build on this very awesome piece of software in order to deliver block-based Arduino programming to the school system where I live. Fingers crossed :)

**Comment by noopkat** [comment]
hi @joachimhs! Thanks for raising this issue. Given that you're looking to integrate avrgirl for Arduino support, have you checked out this [example Chrome app repo](https://github.com/noopkat/avrgirl-chrome-app) for steps on getting all the dependencies working as expected within a Chrome app? Let me know if it works or not or if something doesn't make sense. Thanks, and stoked you're considering avrgirl for your code block app!

**Comment by joachimhs** [comment]
Thank you for the link. I will take a closer look at it. 

I've used webpack and gulp to handle the dependencies. I've also gone a slightly different route, as I already have the block-based web application up and running (Blockuino.no). So the workflow will be 

1. Create the code via drag-and-drop (blockly) (or in the future also write it with text)
2. Click the "upload to Arduino" button. This will then send the source code to the Blockuino backend where it will get compiled using PlatformIO, before returning the .hex file to the web-app
3. The web-app will then message the Chrome app with the hex file, which will upload it to the Arduino. 

I am hoping to support the Uno, Mini and Mega, as well as some of the ATTiny45/85 based micro controllers as a start. 

At some point I will also probably release an English version as well. The main goal is to make Arduino programming accessible for kids down to 4th grade.

**Comment by noopkat** [comment]
Makes sense. The key feature that browserify does is shimming all dependencies for ones that will work in the browser. Webpack doesn't have full shim support for every native Node API, which you can see here: https://github.com/webpack/node-libs-browser/blob/master/index.js. Regardless, it seems you got it working by eliminating localStorage calls. Could you let me know which deps use local storage? Thanks!

**Comment by joachimhs** [comment]
https://www.dropbox.com/s/m17if8j9xkip5ia/Screenshot%202017-02-15%2019.18.44.png?dl=0

I also had to make a small change to avrgirl. This is due to the fact that my hex file-content is stored in a string (which avrgirl assumed was a filename). I brute-force-changed it for now, but I will make it work with both files and strings before submitting a PR

**Comment by noopkat** [comment]
Just an FYI you can convert the string to an array buffer in the browser and avrgirl will accept that with no complaints. The example of this is in the example repo I linked :)

**Comment by joachimhs** [comment]
You are correct, of course :) I will try to use browserify instead of webpack and see if that solves the dependency issue with local storage.

**Comment by joachimhs** [comment]
I don't mean to be spamming you. I have committed a very early working version of the Chrome app. It now uses Browserify to bundle the JavaScript. It do still show the error message that localStorage is not available, but for some reason it also seems to be working fine :) 

It's currently only available in a separate branch of Blockuino :) 
https://github.com/joachimhs/blockuino/tree/blockuino_pi_beta/blockuino-chome-app

**Comment by joachimhs** [comment]
So, I've made some progress here. I've published a very early version of the plugin to the Chrome App Store, so the app is available for Chromebooks. Sadly the Serial-api is not brought along to Chrome Extensions, which means that I will be building something else for win, mac and linux machines, probably with the same stack, only using Electron to build a native application to handle the Serial-communication. 

The Chrome app is available here in case you are interested: https://chrome.google.com/webstore/detail/blockuino-hex-uploader-av/ekbpmcpbckbdpbjhdchoniihflnmabie

I've documented this in a very short (and very Norwegian) video as well :) https://kodegenet.no/figurer/Video.MOV
